browser-extension/mv3/src/service-worker/services/webRequestInterceptor.ts

```diff
@@ -0,0 +1,105 @@
+import { getEnabledRules } from "common/rulesStore";
+import { RuleType } from "common/types";
+import { matchSourceUrl } from "./ruleMatcher";
+import ruleExecutionHandler from "./ruleExecutionHandler";
+
+const onBeforeRequest = (details: chrome.webRequest.WebRequestBodyDetails) => {
+  // @ts-ignore
+  if (details?.documentLifecycle !== "active") {
+    return;
+  }
+
+  getEnabledRules().then((enabledRules) => {
+    let isMainFrameRequest = details.type === "main_frame" ? true : false;
+
+    enabledRules.forEach((rule) => {
+      switch (rule.ruleType) {
+        case RuleType.REDIRECT:
+        case RuleType.REPLACE:
+        case RuleType.QUERYPARAM:
+        case RuleType.CANCEL:
+        case RuleType.DELAY:
+          rule.pairs.forEach((pair) => {
+            if (matchSourceUrl(pair.source, details.url)) {
```

Comment:
Only URL is being matched. We would also need to match source filters: resourceType, requestMethod and pageDomain/initiator domain